### PR TITLE
Creating a Custom Data Transformer using Scikit-Learn

### LINKED ISSUE #2374: Creating a Custom Data Transformer using Scikit-Learn
### Proposed title of the article
Creating a Custom Data Transformer using Scikit-Learn

### Introduction paragraph (2-3 paragraphs):
A Transformer is used to make data fit for the machine learning process while Scikit-Learn is a Python library used for machine learning. It provides built-in methods for data preparation before the data is fed into a training model. However, as a data scientist, you may need further perform custom cleanup processes or adding more attributes that may improve your model's performance. To do that, you will need to create your own custom transformer for your data.
We will look at how to do that using a class, perform feature scaling and then implement the transformer in a pipeline.

### Key takeaways: 
We will learn about:
1. Getting the data
2. Creating a Custom Transformer
3. Transformation Pipelines
4. Handling text and categorical attributes

### References: 
**Book** : Hands-On Machine Learning with Scikit-Learn & TensorFlow CONCEPTS, TOOLS, AND TECHNIQUES TO BUILD INTELLIGENT SYSTEMS by Aurélien Géron

### PR DESCRIPTION
**Pull request by Agusioma**
### Article checklist

**ATTENTION: In your PR - add the ISSUE (topic form) # that this PR corresponds to. You can do this by also adding a comment that states **["This fixes #(enter your Idea Content Form #)"](https://youtu.be/TKJ4RdhyB5Y?t=278)** to link and close your Idea Content (issue) when the PR is merged.**

**NOTE: (Please ensure that you have only one open issue + linked pull request at a time. This will ensure that we complete the article in a timely manner from inception to publishing.)**

**If you have not already please go over our [Resources Page](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md) for more tips and tricks on how to improve our overall technical writing.**

#### Preliminary Checks - Formatting and Structure
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**
- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**
- [x] Have you included a hero image that is 600x400 pixels and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
- [x] In your PR - add a comment that states **["This fixes #(enter your Idea Content Form #)"](https://youtu.be/TKJ4RdhyB5Y?t=278)** to link and close your Idea Content (issue) when the PR is merged.

#### Grammar & plagiarism checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) -  [Grammarly](https://grammarly.com)
- [x] Please place your article through a 3rd party plagiarism checker? We suggest using Quetext, [this tool](https://www.quetext.com/) is free and has a daily limit. **We typically accept articles with 10% or less.**
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [x] Have you added sources for quotes and images that aren't yours?

#### Technical checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**

### Contribution guidelines
For first-time contributors and for more details, **see [Contributing Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**

This fixes #2374

### CODE REVIEW

**Comment by hectorkambow** [comment]
@srishilesh Please let someone else review this article - I believe you have more than 2-3 articles in reivew 👍  

Thank you

**Comment by srishilesh** [comment]
> @srishilesh Please let someone else review this article - I believe you have more than 2-3 articles in reivew 👍
> 
> Thank you

Sure.

**Comment by hectorkambow** [comment]
plagiarism check done

**Review by lalith1403 — CHANGES_REQUESTED** [review]
Can you also include a Jupyter notebook with the code separated in cells. It will help the reader test and experiment with the concepts. Use Google Colab.

**Comment by Agusioma** [comment]
@lalith1403 the link to the Google Colab has been added.

File: content/articles/custom-transformer/index.md [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+images:
+
+  -url: /engineering-education/custom-transformer/hero.jpg
+   alt: First Hop Redundancy Protocol example image
+---
+
+
+### Introduction
+
+In machine learning, a data transformer is used to make a dataset fit for the training process. Scikit-Learn is a Python library containing methods that help us achieve our machine learning process faster. Scikit-Learn provides built-in methods for data preparation before the data is fed into a training model. However, as a data scientist, you may need to perform more custom cleanup processes or adding more attributes that may improve your model's performance. To do that, you will need to create a custom transformer for your data.
```
Comment (lalith1403) [inline content/articles/custom-transformer/index.md:21]: Scikit learn enable quick experimentation to achieve quality results with a minimal amount of time spent on implementing data pipelines involving preprocessing, machine learning algorithms, evaluation and inference.

Change the second sentence.

File: content/articles/custom-transformer/index.md [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+```
+
+#### Cleaning the data
+
+The cleaning operation we will do here is filling empty numeric attributes with their median values. We will use the `SimpleImputer`, an estimator, to do that.
+
+```python 
+from sklearn.impute import SimpleImputer
+
+imputer = SimpleImputer(strategy="median")
```
Comment (lalith1403) [inline content/articles/custom-transformer/index.md:101]: Explain the code line by line. What is median strategy?

Comment (lalith1403) [reply content/articles/custom-transformer/index.md:101]: If the line is obvious, then there is no need to explain.

File: content/articles/custom-transformer/index.md [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+
+We drop the *ocean_proximity* attribute because it's a text attribute that will handle in the next section.
+
+The result produced is an array, so we convert it to a DataFrame.
+
+#### Handling text and categorical attributes
+
+We cannot handle text and numerical attributes similarly. For example, we cannot compute the median of text.
+
+We will use a transformer for this called the `OrdinalEncoder`. `OrdinalEncoder` is chosen because it is more pipeline friendly.
```
Comment (lalith1403) [inline content/articles/custom-transformer/index.md:118]: What does ordinalencoder do?

File: content/articles/custom-transformer/index.md [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+```python
+from sklearn.preprocessing import OrdinalEncoder
+
+our_text_cats = our_dataset[['ocean_proximity']]
+our_encoder = OrdinalEncoder()
+our_encoded_dataset = our_encoder.fit_transform(our_text_cats)
+
+```
+
+### Our transformer
```
Comment (lalith1403) [inline content/articles/custom-transformer/index.md:129]: Data transformer. Transformers have different meanings under machine learning. It's better to be precise.

File: content/articles/custom-transformer/index.md [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+
+```python
+
+import numpy as np
+from sklearn.base import BaseEstimator, TransformerMixin
+#initialising column numbers
+rooms,  bedrooms, population, household = 3,4,5,6
+
+class CustomTransformer(BaseEstimator, TransformerMixin):
+    def __init__(self, add_bedrooms_per_room = True):
```
Comment (lalith1403) [inline content/articles/custom-transformer/index.md:144]: Explain this class. How did you start to approach the logic? How did you break it into different methods? Explain the thought process step by step.

**Review by lalith1403 — CHANGES_REQUESTED** [review]
A good article explaining the basics of data transformers and data pipelines. I would suggest add more explanation regarding the thought process behind each line of code.

**Comment by Agusioma** [comment]
@lalith1403 I fixed some changes. check

**Comment by Agusioma** [comment]
@lalith1403

**Comment by ahmadmardeni1** [comment]
I can't access your [Google Colab notebook](https://drive.google.com/file/u/0/d/1DVIh0LhGOU0rwVU2bbZw_VmjiXWV-daM/edit).
It is asking for a password, please change the setting so that anyone with the link can access it.
Will merge when you do that. @Agusioma

**Comment by ahmadmardeni1** [comment]
I can't access your [Google Colab notebook](https://drive.google.com/file/u/0/d/1DVIh0LhGOU0rwVU2bbZw_VmjiXWV-daM/edit).
It is asking for a password, please change the settings so that anyone with the link can access it.
Will merge when you do that. @Agusioma

**Comment by ahmadmardeni1** [comment]
Good work gentlemen @Agusioma @lalith1403

**Pull request merged**